frontend/src/app/main/data/events.cljs

```diff
@@ -0,0 +1,238 @@
+
+(defn- event->generic-action
+  [event name]
+  {:type "action"
+   :name name
+   :props {}})
+
+(defmethod process-event :app.main.data.users/logout
+  [event]
+  (event->generic-action event "logout"))
+
+
+;; --- MAIN LOOP
+
+(defn- append-to-buffer
+  [buffer item]
+  (if (>= (count buffer) max-buffer-size)
+    buffer
+    (conj buffer item)))
+
+(defn- remove-from-buffer
+  [buffer items]
+  (into #queue [] (drop items) buffer))
+
+(defn- persist-events
+  [events]
+  (let [uri    (u/join cf/public-uri "events")
+        params {:events events}]
+    (->> (http/send! {:uri uri
+                      :method :post
+                      :body (http/transit-data params)})
+         (rx/mapcat rp/handle-response))))
+
+(defmethod ptk/resolve ::persistence
+  [_ {:keys [buffer] :as params}]
+  (ptk/reify ::persistence
+    ptk/EffectEvent
+    (effect [_ state stream]
+      (let [events (into [] (take max-chunk-size) @buffer)]
+        (when-not (empty? events)
+          (->> (persist-events events)
+               (rx/subs (fn [_]
+                          (swap! buffer remove-from-buffer (count events))))))))))
+
+(defn initialize
+  []
+  (let [buffer (atom #queue [])]
+    (ptk/reify ::initialize
+      ptk/WatchEvent
+      (watch [_ state stream]
+        (->> (rx/from-atom buffer)
+             (rx/filter #(pos? (count %)))
+             (rx/debounce 2000)
+             (rx/map #(ptk/event ::persistence {:buffer buffer}))))
+
+      ptk/EffectEvent
+      (effect [_ state stream]
+        (let [events  (methods process-event)
+              session (atom nil)
+
+              profile (->> (rx/from-atom storage {:emit-current-value? true})
+                           (rx/map :profile)
+                           (rx/map :id)
+                           (rx/dedupe))
+
+              source  (->> stream
+                           (rx/with-latest-from profile)
+                           (rx/map (fn [result]
+                                     (let [event      (aget result 0)
+                                           profile-id (aget result 1)
+                                           type       (ptk/type event)
+                                           impl-fn    (get events type)]
+                                       (when (fn? impl-fn)
+                                         (some-> (impl-fn event)
+                                                 (update :profile-id #(or % profile-id)))))))
+                           (rx/filter :profile-id)
+                           (rx/map (fn [event]
+                                     (let [session* (or @session (dt/now))
+                                           context  (-> @context
+                                                        (d/merge (:context event))
+                                                        (assoc :session session*))]
+                                       (swap! session (constantly session*))
```

Comment:
Could it be `(reset! session session*)`?

Reply:
you are right, the swap! here is unnecesary. I'll change it.